package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+                                                        dtype=np.float32)
+                self._set_attr_unit(self._time, 'time')
+                break
+
+    def _create_trajectory_dataset(self, group):
+        """helper function to initialize a dataset for
+        position, velocity, and force"""
+
+        if self.n_frames is None:
+            shape = (0, self.n_atoms, 3)
+            maxshape = (None, self.n_atoms, 3)
+        else:
+            shape = (self.n_frames, self.n_atoms, 3)
+            maxshape = None
+
+        chunks = None if self.contiguous else self.chunks
+
+        self._traj.require_group(group)
+        self._traj.require_dataset(f'{group}/value',
+                                   shape=shape,
+                                   maxshape=maxshape,
+                                   dtype=np.float32,
+                                   chunks=chunks,
+                                   compression=self.compression,
+                                   compression_opts=self.compression_opts)
+        if 'step' not in self._traj[group]:
+            self._traj[f'{group}/step'] = self._step
+        if 'time' not in self._traj[group]:
+            self._traj[f'{group}/time'] = self._time
+
+    def _create_observables_dataset(self, group, data):
+        """helper function to initialize a dataset for each observable"""
+
+        self._obsv.require_group(group)
+        # guarantee ints and floats have a shape ()
+        data = np.asarray(data)
+        self._obsv.require_dataset(f'{group}/value',
+                                   shape=(0,) + data.shape,
+                                   maxshape=(None,) + data.shape,
+                                   dtype=data.dtype)
+        if 'step' not in self._obsv[group]:
+            self._obsv[f'{group}/step'] = self._step
+        if 'time' not in self._obsv[group]:
+            self._obsv[f'{group}/time'] = self._time
+
+    def _set_attr_unit(self, dset, unit):
+        """helper function to set a 'unit' attribute for an HDF5 dataset"""
+
+        if self.units[unit] is None:
+            return
+
+        dset.attrs['unit'] = self._unit_translation_dict[unit][self.units[unit]]
+
+    def _write_next_timestep(self, ts):
+        """Write coordinates and unitcell information to H5MD file.
+
+        Do not call this method directly; instead use
+        :meth:`write` because some essential setup is done
+        there before writing the first frame.
+
+        The first dimension of each dataset is extended by +1 and
+        then the data is written to the new slot.
+
+        Note
+        ----
+        Writing H5MD files with fancy trajectory slicing where the Timestep
+        does not increase monotonically such as ``u.trajectory[[2,1,0]]``
+        or ``u.trajectory[[0,1,2,0,1,2]]`` raises a :exc:`ValueError` as this
+        violates the rules of the step dataset in the H5MD standard.
+
+        """
+
+        i = self._counter
```

Comment:
A more descriptive counter here might be worth it. I'm not in the "i,j,k,l are terrible" camp, but at the same time if they get created outside of a loop it's really easy for someone to accidentally do `for i in range(N)` somewhere in your code and forget that you had `i` already declared somewhere else.

Reply:
I think `i` is fine here because it makes the rest of the code more readable.

Reply:
I like `i` since it's always used as a counter, but I see your point in that since _it's always used as a counter_ someone might accidently overwrite it. I'll leave it for now?